third_party/move/move-vm/runtime/src/data_cache.rs

```diff
@@ -234,15 +380,27 @@ impl TransactionDataCache {
         addr: &AccountAddress,
         ty: &Type,
     ) -> PartialVMResult<&mut GlobalValue> {
-        if let Some(account_cache) = self.account_map.get_mut(addr) {
-            if let Some(entry) = account_cache.get_mut(ty) {
-                return Ok(&mut entry.value);
-            }
+        if let Some(entry) = self.find_entry_mut(addr, ty) {
+            return entry.value.value_mut();
         }
 
         let msg = format!("Resource for {:?} at {} must exist", ty, addr);
         let err =
             PartialVMError::new(StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR).with_message(msg);
         Err(err)
     }
+
+    pub(crate) fn get_resource_existence(
+        &mut self,
+        addr: &AccountAddress,
+        ty: &Type,
+    ) -> PartialVMResult<bool> {
+        if let Some(entry) = self.find_entry_mut(addr, ty) {
+            return entry.value.exists();
+        }
+
+        let msg = format!("Resource for {:?} at {} must exist", ty, addr);
+        let err = PartialVMError::new_invariant_violation(msg);
+        Err(err)
```

Comment:
nit: I think this can all be embedded, or even better we can just do (with better formatting)
``
self.find_entry_mut(addr, ty).map(|entry| entry.value.exists()).ok_or_else(|| PartialVMError::new_invariant_violation(format!("Entry for resource {:?} at {} must exist", ty, addr))
``

or even better, since I see get_resource_mut does a similar pattern.
We can just adjust find_entry_mut to be expect_entry_mut (assuming there are other use cases that may not find an entry) and embed the error inside there already. Then we can just do the map
`self.expect_entry_mut(addr, ty).map(|entry| entry.value.exists())` here and in get_resource_mut as well